kamu-core/src/infra/sync_service_impl.rs

```diff
@@ -268,6 +275,73 @@ impl SyncServiceImpl {
         }
     }
 
+    async fn sync_smart_pull_transfer_protocol<'a>(
+        &'a self,
+        odf_src: &DatasetRefAny,
+        dst: &DatasetRefAny,
+        opts: SyncOptions,
+        listener: Arc<dyn SyncListener>,
+    ) -> Result<SyncResult, SyncError> {
+        let odf_src_remote_ref = odf_src.as_remote_ref().unwrap();
+        let odf_src_url = self.resolve_remote_dataset_url(&odf_src_remote_ref).await?;
+        let http_src_url = Url::parse(&(odf_src_url.as_str())[4..]).unwrap(); // odf+http, odf+https - cut odf+
+
+        let dst_dataset_builder = self
+            .get_dataset_writer(dst, opts.create_if_not_exists)
+            .await?;
+
+        let dst_dataset = dst_dataset_builder.as_dataset();
+
+        info!("Starting sync using Smart Transfer Protocol (Pull flow)");
+        self.smart_transfer_protocol
+            .pull_protocol_client_flow(&http_src_url, dst_dataset, listener)
+            .await
+    }
+
+    async fn sync_smart_push_transfer_protocol<'a>(
+        &'a self,
+        src: &DatasetRefAny,
+        odf_dst: &DatasetRefAny,
+        listener: Arc<dyn SyncListener>,
+    ) -> Result<SyncResult, SyncError> {
+        let src_dataset = self.get_dataset_reader(src).await?;
+
+        let odf_dst_remote_ref = odf_dst.as_remote_ref().unwrap();
+        let odf_dst_url = self.resolve_remote_dataset_url(&odf_dst_remote_ref).await?;
+        let http_dst_url = Url::parse(&(odf_dst_url.as_str())[4..]).unwrap(); // odf+http, odf+https - cut odf+
+
+        info!("Starting sync using Smart Transfer Protocol (Push flow)");
+        self.smart_transfer_protocol
+            .push_protocol_client_flow(src_dataset.as_ref(), &http_dst_url, listener)
+            .await
+    }
+
+    async fn sync_simple_transfer_protocol<'a>(
+        &'a self,
+        src: &'a dyn Dataset,
+        src_ref: &'a DatasetRefAny,
+        dst: &'a dyn Dataset,
+        _dst_ref: &'a DatasetRefAny,
+        validation: AppendValidation,
+        trust_source_hashes: bool,
+        force: bool,
+        listener: Arc<dyn SyncListener + 'static>,
+    ) -> Result<SyncResult, SyncError> {
+        info!("Starting sync using Simple Transfer Protocol");
+        SimpleTransferProtocol
+            .sync(
+                src,
+                src_ref,
+                dst,
+                _dst_ref,
```

Comment:
underscore should be removed since variable is in use?

Reply:
oops